### PR TITLE
Create uploader actor

### PR DESCRIPTION
**Pull request by Ruadhri17**
## Proposed changes
This PR adds new uploader actor that works in similar manner to `downloader_actor`. Later on it will be used in both `tedge-log-plugin` and `tedge-configuration-plugin` to take responsibility for uploading content to tedge file repository. More in #2301.

TODO:

- [x] Add test that checks if backoff strategy is working.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
* #2301
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by albinsuresh — COMMENTED** [review]
The overall code looks fine with the basic retry mechanism. But, unlike the downloader which could resume a download from the point where it was interrupted, I didn't see a similar resumption mechanism here. Is there any plan to add such a multi-part upload functionality to this uploader? It doesn't necessarily have to be in this PR, but would be a good follow-up addition.

BTW, I didn't see the log plugin or config plugin using this actor. Isn't that in scope either?

File: crates/common/upload/src/upload.rs
Comment (rina23q) [inline crates/common/upload/src/upload.rs:196]: You can also create a test `upload_content_with_auth()` based on this test.
The change will be only below:

```rust
let _mock1 = server
    .mock("PUT", "/some_file.txt")
    .with_status(201)
    .match_header("Authorization", mockito::Matcher::Regex(r"Bearer .*".to_string()))
    .create();

let url = UploadInfo::new(&target_url).with_auth(Auth::Bearer("1234".to_string()));

```

**Comment by Ruadhri17** [comment]
> The overall code looks fine with the basic retry mechanism. But, unlike the downloader which could resume a download from the point where it was interrupted, I didn't see a similar resumption mechanism here. Is there any plan to add such a multi-part upload functionality to this uploader? It doesn't necessarily have to be in this PR, but would be a good follow-up addition.

I would leave the partial upload for next PR as it is more tricky than download. According to the [rfc-editor](https://www.rfc-editor.org/rfc/rfc9110.html#name-partial-put) this is very inconsistent and require some changes on server side. Another thing is that we would have to edit the exponential backoff mechanism because of 400 BAD REQUEST respond which at this point will not trigger retry mechanism.

File: crates/common/upload/src/upload.rs
Comment (Bravo555) [inline crates/common/upload/src/upload.rs:110]: question: Do we always want to set it to `application/octet-stream` though? I realize that it's a safe default and then it's the problem of who received that data to figure out what MIME is it, but maybe we can use a crate for detecting MIME type and then use that.
But it's not a priority and could be somewhat complex, so if you agree, then it could be done in another PR some other time.

File: crates/extensions/tedge_log_manager/src/actor.rs
Comment (Bravo555) [inline crates/extensions/tedge_log_manager/src/actor.rs:201]: issue: How do we know topic passed here is always valid? I'd rather we use `Topic::new()` and return the error, or take `topic: &Topic` as an argument.

File: crates/extensions/tedge_log_manager/src/actor.rs
Comment (Bravo555) [inline crates/extensions/tedge_log_manager/src/actor.rs:34]: suggestion: Not all strings are valid MQTT topics, so I'd rather see a type like `Topic` used here if possible.

**Review by Bravo555 — CHANGES_REQUESTED** [review]

File: crates/extensions/tedge_config_manager/src/tests.rs
Comment (albinsuresh) [inline crates/extensions/tedge_config_manager/src/tests.rs:208]:
```suggestion
    // Simulate upload file completion
```

File: crates/common/log_manager/src/log_utils.rs
Comment (albinsuresh) [inline crates/common/log_manager/src/log_utils.rs:1]: A library returning `anyhow` errors is highly discouraged. You should be defining concrete error types with the use of `thiserror` instead. Anyhow is usually used at the application level where the errors just need to be reported.

File: crates/common/log_manager/src/log_utils.rs
Comment (albinsuresh) [inline crates/common/log_manager/src/log_utils.rs:163]:
```suggestion
fn temp_file_at(path: &Path) -> Result<PathBuf, LogRetrievalError> {
```

File: crates/common/log_manager/src/log_utils.rs
Comment (albinsuresh) [inline crates/common/log_manager/src/log_utils.rs:168]: I see that this forced you to use `anyhow` as a dependency. Using `anyhow` in the `source` field of `FileError::InvalidFileName` was a mistake. As a short term solution, I'd suggest that you use `PathsError` here as well. Feel free to add one `InvalidFileName` variant as well.

**Review by albinsuresh — COMMENTED** [review]
The main changes look fine. Just some minor niggles and I'm happy to approve once those are ironed out.

File: crates/common/log_manager/src/log_utils.rs
Comment (rina23q) [inline crates/common/log_manager/src/log_utils.rs:168]: Why dou you need `anyhow!`?

File: crates/extensions/tedge_log_manager/src/actor.rs
Comment (rina23q) [reply crates/extensions/tedge_log_manager/src/actor.rs:201]: The HashMap `self.pending_operations` keep topic names (String) that are waiting for uploads by the upload actor. So, the topic from the hashmap should be always valid.

**Review by rina23q — COMMENTED** [review]
Except `anyhow` in a library crate, looks good to me

File: crates/common/log_manager/src/log_utils.rs
Comment (Bravo555) [reply crates/common/log_manager/src/log_utils.rs:1]: IMO there are valid use cases for using `anyhow` in a library crate. What matters the most is the intent, what do we expect users to do. We do not expect code which calls the uploader to behave differently between error variants of e.g. "this does not name a valid file" or "path is not valid unicode", we simply return these errors and show them to the user. We don't have to maintain any Rust API compatibility, we can add or remove error variants at any time.
So I think in this case using an error variant for "file error" and then using `anyhow` for telling the user what exactly went wrong is acceptable.
`anyhow::Error` is also like `Box<dyn Error> + 'static`, but also guarantees `Send` and `Sync`. I originally added it to `InvalidFileName` error variant because i wanted to use either `io::Error` or `Utf8Error` as a source, but using `anyhow!("error details")` directly... I wouldn't say it's an antipattern.

I was recently reading _Zero To Production In Rust_ and there's a free chapter about error handling. There is a dedicated [section about anyhow vs thiserror](https://www.lpalmieri.com/posts/error-handling-rust/#anyhow-or-thiserror), but I'd recommend reading the entire chapter for context.

File: crates/common/log_manager/src/log_utils.rs
Comment (Bravo555) [reply crates/common/log_manager/src/log_utils.rs:168]: anyhow::Error is also like Box<dyn Error> + 'static, but also guarantees Send and Sync. I originally added it to InvalidFileName error variant because i wanted to use either io::Error or Utf8Error as a source, but I wouldn't say using `anyhow!` here to provide a a description of why is the filename invalid, to the user, is an antipattern. This could've just been a string, but because it's `anyhow::Error`, one can also use any other error as a source.
I addressed creating more enum variants in this comment: https://github.com/thin-edge/thin-edge.io/pull/2302#discussion_r1354377885.

File: crates/common/log_manager/src/log_utils.rs
Comment (albinsuresh) [reply crates/common/log_manager/src/log_utils.rs:1]: Thanks for the link. Gave me a new perspective. I guess the following is the key point:

```
Library authors cannot (or do not want to) make assumptions on the intent of their users. They steer away from being opinionated (to an extent) - enums give users more control, if they need it.
```

and I completely agree to your following point:

> We do not expect code which calls the uploader to behave differently between error variants of e.g. "this does not name a valid file" or "path is not valid unicode", we simply return these errors and show them to the user.

Here is where I'm wondering what value `anyhow` adds if all we wanted to do was to report the reason to the end-user. Why couldn't it be just a `String` value? Using `anyhow` instead, is just an overkill, right? As the user is less likely to introspect the `source` chain any further, as that was our fundamental assumption when we picked `anyhow` for "opaque" errors in the first place.

File: crates/common/upload/src/upload.rs
Comment (albinsuresh) [reply crates/common/upload/src/upload.rs:110]: Good point. @Ruadhri17 could you please confirm that the C8y UI is able to properly render the uploaded files as text files in the UI? Else we have a problem and might need to implement some "file content type detection" before setting this.

Comment (Ruadhri17) [reply crates/common/upload/src/upload.rs:110]: As there is plan to introduce partial PUT, we can also add MIME type detection then. But I agree: it would be much better to  provide type of data we are sending to the server.

Comment (Ruadhri17) [reply crates/common/upload/src/upload.rs:110]: > Good point. @Ruadhri17 could you please confirm that the C8y UI is able to properly render the uploaded files as text files in the UI? Else we have a problem and might need to implement some "file content type detection" before setting this.

Sure I will test it and if there will be no issues I will request for a review.

Comment (Bravo555) [reply crates/common/upload/src/upload.rs:110]: @Ruadhri17 confirmed that there are no issues with rendering uploaded files in Cumulocity UI.

**Review by Bravo555 — APPROVED** [review]

**Review by albinsuresh — APPROVED** [review]
I still have my reservations against the use of `anyhow` to represent a source in error variants like `InvalidFileName` where that enum itself was self-explanatory and I doubt if the `source` would really add any more context. But, that's a parallel discussion that should not block this PR. Hence approving as the rest of the code looks fine,

**Comment by gligorisaev** [comment]
QA has thoroughly checked the feature and here are the results:

- [ ] Test for ticket exists in the test suite.
- [x] QA has tested the bug and could not reproduce it anymore.

**Pull request merged**
